// astronomy-bundle/earth/Earth.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var AstronomicalObject_1 = require("../astronomicalObject/AstronomicalObject");
var utils_1 = require("../utils");
var calculationCache_1 = require("../cache/calculationCache");
var vsop87Calc_1 = require("../utils/vsop87Calc");
var angleCalc_1 = require("../utils/angleCalc");
var coordinateCalc_1 = require("../utils/coordinateCalc");
var Earth = (function (_super) {
    __extends(Earth, _super);
    function Earth() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'earth';
        return _this;
    }
    Earth.prototype.getHeliocentricEclipticRectangularJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getHeliocentricEclipticSphericalJ2000Coordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.spherical2rectangular(coords)];
                }
            });
        });
    };
    Earth.prototype.getHeliocentricEclipticRectangularDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getHeliocentricEclipticSphericalDateCoordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.spherical2rectangular(coords)];
                }
            });
        });
    };
    Earth.prototype.getHeliocentricEclipticSphericalJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, calculationCache_1.getAsyncCachedCalculation('earth_heliocentric_spherical_j2000', this.t, function () { return __awaiter(_this, void 0, void 0, function () {
                            var vsop87;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4, Promise.resolve().then(function () { return require('./vsop87/vsop87EarthSphericalJ2000'); })];
                                    case 1:
                                        vsop87 = _a.sent();
                                        return [2, {
                                                lon: angleCalc_1.normalizeAngle(vsop87Calc_1.calculateVSOP87Angle(vsop87.VSOP87_X, this.t)),
                                                lat: vsop87Calc_1.calculateVSOP87Angle(vsop87.VSOP87_Y, this.t),
                                                radiusVector: vsop87Calc_1.calculateVSOP87(vsop87.VSOP87_Z, this.t),
                                            }];
                                }
                            });
                        }); })];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Earth.prototype.getHeliocentricEclipticSphericalDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, calculationCache_1.getAsyncCachedCalculation('earth_heliocentric_spherical_date', this.t, function () { return __awaiter(_this, void 0, void 0, function () {
                            var vsop87;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4, Promise.resolve().then(function () { return require('./vsop87/vsop87EarthSphericalDate'); })];
                                    case 1:
                                        vsop87 = _a.sent();
                                        return [2, {
                                                lon: angleCalc_1.normalizeAngle(vsop87Calc_1.calculateVSOP87Angle(vsop87.VSOP87_X, this.t)),
                                                lat: vsop87Calc_1.calculateVSOP87Angle(vsop87.VSOP87_Y, this.t),
                                                radiusVector: vsop87Calc_1.calculateVSOP87(vsop87.VSOP87_Z, this.t),
                                            }];
                                }
                            });
                        }); })];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Earth.prototype.getGeocentricEclipticRectangularJ2000Coordinates = function () {
        return Promise.reject({ x: 0, y: 0, z: 0 });
    };
    Earth.prototype.getGeocentricEclipticRectangularDateCoordinates = function () {
        return Promise.reject({ x: 0, y: 0, z: 0 });
    };
    Earth.prototype.getGeocentricEclipticSphericalJ2000Coordinates = function () {
        return Promise.reject({ lon: 0, lat: 0, radiusVector: 0 });
    };
    Earth.prototype.getGeocentricEclipticSphericalDateCoordinates = function () {
        return Promise.reject({ lon: 0, lat: 0, radiusVector: 0 });
    };
    Earth.prototype.getApparentGeocentricEclipticSphericalCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Promise.reject({ lon: 0, lat: 0, radiusVector: 0 })];
            });
        });
    };
    Earth.prototype.getNutationInLongitude = function () {
        return utils_1.earthCalc.getNutationInLongitude(this.T);
    };
    Earth.prototype.getNutationInObliquity = function () {
        return utils_1.earthCalc.getNutationInObliquity(this.T);
    };
    Earth.prototype.getMeanObliquityOfEcliptic = function () {
        return utils_1.earthCalc.getMeanObliquityOfEcliptic(this.T);
    };
    Earth.prototype.getTrueObliquityOfEcliptic = function () {
        return utils_1.earthCalc.getTrueObliquityOfEcliptic(this.T);
    };
    return Earth;
}(AstronomicalObject_1.default));
exports.default = Earth;
