// astronomy-bundle/stars/Star.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var AstronomicalObject_1 = require("../astronomicalObject/AstronomicalObject");
var epoch_1 = require("../constants/epoch");
var coordinateCalc_1 = require("../utils/coordinateCalc");
var starCalc_1 = require("../utils/starCalc");
var precessionCalc_1 = require("../utils/precessionCalc");
var apparentCoordinateCalc_1 = require("../utils/apparentCoordinateCalc");
var Star = (function (_super) {
    __extends(Star, _super);
    function Star(equatorialCoords, toi, properMotion, referenceEpoch) {
        if (properMotion === void 0) { properMotion = { rightAscension: 0, declination: 0 }; }
        if (referenceEpoch === void 0) { referenceEpoch = epoch_1.EPOCH_J2000; }
        var _this = _super.call(this, toi) || this;
        _this.equatorialCoords = equatorialCoords;
        _this.properMotion = properMotion;
        _this.referenceEpoch = referenceEpoch;
        return _this;
    }
    Star.prototype.getHeliocentricEclipticSphericalJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticSphericalJ2000Coordinates()];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Star.prototype.getHeliocentricEclipticSphericalDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticSphericalDateCoordinates()];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Star.prototype.getHeliocentricEclipticRectangularJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticRectangularJ2000Coordinates()];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Star.prototype.getHeliocentricEclipticRectangularDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticRectangularDateCoordinates()];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    Star.prototype.getGeocentricEclipticRectangularJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticSphericalJ2000Coordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.spherical2rectangular(coords)];
                }
            });
        });
    };
    Star.prototype.getGeocentricEclipticRectangularDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticSphericalDateCoordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.spherical2rectangular(coords)];
                }
            });
        });
    };
    Star.prototype.getGeocentricEclipticSphericalJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEquatorialSphericalJ2000Coordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.equatorialSpherical2eclipticSpherical(coords, this.T)];
                }
            });
        });
    };
    Star.prototype.getGeocentricEclipticSphericalDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEquatorialSphericalDateCoordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.equatorialSpherical2eclipticSpherical(coords, this.T)];
                }
            });
        });
    };
    Star.prototype.getGeocentricEquatorialSphericalJ2000Coordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Promise.resolve(this.equatorialCoords)];
            });
        });
    };
    Star.prototype.getGeocentricEquatorialSphericalDateCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                coords = starCalc_1.correctProperMotion(this.equatorialCoords, this.properMotion, this.jd, this.referenceEpoch);
                return [2, Promise.resolve(precessionCalc_1.correctPrecessionForEquatorialCoordinates(coords, this.jd, this.referenceEpoch))];
            });
        });
    };
    Star.prototype.getApparentGeocentricEclipticSphericalCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getGeocentricEclipticSphericalDateCoordinates()];
                    case 1:
                        coords = _a.sent();
                        coords = apparentCoordinateCalc_1.correctEffectOfAberration(coords, this.T);
                        return [2, apparentCoordinateCalc_1.correctEffectOfNutation(coords, this.T)];
                }
            });
        });
    };
    Star.prototype.getApparentGeocentricEquatorialSphericalCoordinates = function () {
        return __awaiter(this, void 0, void 0, function () {
            var coords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getApparentGeocentricEclipticSphericalCoordinates()];
                    case 1:
                        coords = _a.sent();
                        return [2, coordinateCalc_1.eclipticSpherical2equatorialSpherical(coords, this.T)];
                }
            });
        });
    };
    return Star;
}(AstronomicalObject_1.default));
exports.default = Star;
