// astronomy-bundle/utils/moonCalc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOpticalLiberationInLatitude = exports.getOpticalLiberationInLongitude = exports.getDistanceToEarth = exports.getRadiusVector = exports.getLatitude = exports.getLongitude = exports.getEquatorialHorizontalParallax = exports.getMeanLongitudeOfAscendingNode = exports.getMeanLongitude = exports.getArgumentOfLatitude = exports.getMeanAnomaly = exports.getMeanElongation = void 0;
var moon_1 = require("../constants/moon");
var angleCalc_1 = require("./angleCalc");
var distanceCalc_1 = require("./distanceCalc");
var index_1 = require("./index");
function getMeanElongation(T) {
    var D = 297.8501921
        + 445267.1114034 * T
        - 0.0018819 * Math.pow(T, 2)
        + Math.pow(T, 3) / 545868
        - Math.pow(T, 4) / 113065000;
    return angleCalc_1.normalizeAngle(D);
}
exports.getMeanElongation = getMeanElongation;
function getMeanAnomaly(T) {
    var Mmoon = 134.9633964
        + 477198.8675055 * T
        + 0.0087414 * Math.pow(T, 2)
        + Math.pow(T, 3) / 69699
        - Math.pow(T, 4) / 1471200;
    return angleCalc_1.normalizeAngle(Mmoon);
}
exports.getMeanAnomaly = getMeanAnomaly;
function getArgumentOfLatitude(T) {
    var F = 93.2720950
        + 483202.0175233 * T
        - 0.0036539 * Math.pow(T, 2)
        - Math.pow(T, 3) / 352600
        + Math.pow(T, 4) / 86331000;
    return angleCalc_1.normalizeAngle(F);
}
exports.getArgumentOfLatitude = getArgumentOfLatitude;
function getMeanLongitude(T) {
    var L = 218.3164477
        + 481267.88123421 * T
        - 0.0015786 * Math.pow(T, 2)
        + Math.pow(T, 3) / 538841
        - Math.pow(T, 4) / 65194000;
    return angleCalc_1.normalizeAngle(L);
}
exports.getMeanLongitude = getMeanLongitude;
function getMeanLongitudeOfAscendingNode(T) {
    return 125.0445479
        - 1934.1362891 * T
        + 0.0020754 * Math.pow(T, 2)
        + Math.pow(T, 3) / 467441
        - Math.pow(T, 4) / 60616000;
}
exports.getMeanLongitudeOfAscendingNode = getMeanLongitudeOfAscendingNode;
function getEquatorialHorizontalParallax(T) {
    var d = getDistanceToEarth(T);
    return angleCalc_1.rad2deg(Math.asin(6378.14 / d));
}
exports.getEquatorialHorizontalParallax = getEquatorialHorizontalParallax;
function getLongitude(T) {
    var L = getMeanLongitude(T);
    var sumL = _getSumL(T);
    return L + (sumL / 1000000);
}
exports.getLongitude = getLongitude;
function getLatitude(T) {
    var sumB = _getSumB(T);
    return sumB / 1000000;
}
exports.getLatitude = getLatitude;
function getRadiusVector(T) {
    return distanceCalc_1.km2au(getDistanceToEarth(T));
}
exports.getRadiusVector = getRadiusVector;
function getDistanceToEarth(T) {
    var sumR = _getSumR(T);
    return 385000.56 + (sumR / 1000);
}
exports.getDistanceToEarth = getDistanceToEarth;
function _getSumR(T) {
    var D = getMeanElongation(T);
    var Msun = index_1.sunCalc.getMeanAnomaly(T);
    var Mmoon = getMeanAnomaly(T);
    var F = getArgumentOfLatitude(T);
    var E = 1 - 0.002516 * T - 0.0000074 * Math.pow(T, 2);
    var sumR = 0;
    moon_1.MOON_ARGUMENTS_LR.forEach(function (args) {
        var argD = args[0];
        var argMsun = args[1];
        var argMmoon = args[2];
        var argF = args[3];
        var argSumR = args[5];
        var tmpSumR = Math.cos(angleCalc_1.deg2rad(argD * D + argMsun * Msun + argMmoon * Mmoon + argF * F));
        switch (argMsun) {
            case 1:
            case -1:
                tmpSumR = tmpSumR * argSumR * E;
                break;
            case 2:
            case -2:
                tmpSumR = tmpSumR * argSumR * E * E;
                break;
            default:
                tmpSumR = tmpSumR * argSumR;
                break;
        }
        sumR += tmpSumR;
    });
    return sumR;
}
function _getSumL(T) {
    var L = getMeanLongitude(T);
    var D = getMeanElongation(T);
    var Msun = index_1.sunCalc.getMeanAnomaly(T);
    var Mmoon = getMeanAnomaly(T);
    var F = getArgumentOfLatitude(T);
    var A1 = 119.75 + 131.849 * T;
    var A2 = 53.09 + 479264.290 * T;
    var E = 1 - 0.002516 * T - 0.0000074 * Math.pow(T, 2);
    var sumL = 3958 * Math.sin(angleCalc_1.deg2rad(A1))
        + 1962 * Math.sin(angleCalc_1.deg2rad(L - F))
        + 318 * Math.sin(angleCalc_1.deg2rad(A2));
    moon_1.MOON_ARGUMENTS_LR.forEach(function (args) {
        var argD = args[0];
        var argMsun = args[1];
        var argMmoon = args[2];
        var argF = args[3];
        var argSumL = args[4];
        var tmpSumL = Math.sin(angleCalc_1.deg2rad(argD * D + argMsun * Msun + argMmoon * Mmoon + argF * F));
        switch (argMsun) {
            case 1:
            case -1:
                tmpSumL = tmpSumL * argSumL * E;
                break;
            case 2:
            case -2:
                tmpSumL = tmpSumL * argSumL * E * E;
                break;
            default:
                tmpSumL = tmpSumL * argSumL;
                break;
        }
        sumL += tmpSumL;
    });
    return sumL;
}
function _getSumB(T) {
    var L = getMeanLongitude(T);
    var D = getMeanElongation(T);
    var Msun = index_1.sunCalc.getMeanAnomaly(T);
    var Mmoon = getMeanAnomaly(T);
    var F = getArgumentOfLatitude(T);
    var A1 = 119.75 + 131.849 * T;
    var A3 = 313.45 + 481266.484 * T;
    var E = 1 - 0.002516 * T - 0.0000074 * Math.pow(T, 2);
    var sumB = -2235 * Math.sin(angleCalc_1.deg2rad(L))
        + 382 * Math.sin(angleCalc_1.deg2rad(A3))
        + 175 * Math.sin(angleCalc_1.deg2rad(A1 - F))
        + 175 * Math.sin(angleCalc_1.deg2rad(A1 + F))
        + 127 * Math.sin(angleCalc_1.deg2rad(L - Mmoon))
        - 115 * Math.sin(angleCalc_1.deg2rad(L + Mmoon));
    moon_1.MOON_ARGUMENTS_B.forEach(function (args) {
        var argD = args[0];
        var argMsun = args[1];
        var argMmoon = args[2];
        var argF = args[3];
        var argSumB = args[4];
        var tmpSumB = Math.sin(angleCalc_1.deg2rad(argD * D + argMsun * Msun + argMmoon * Mmoon + argF * F));
        switch (argMsun) {
            case 1:
            case -1:
                tmpSumB = tmpSumB * argSumB * E;
                break;
            case 2:
            case -2:
                tmpSumB = tmpSumB * argSumB * Math.pow(E, 2);
                break;
            default:
                tmpSumB = tmpSumB * argSumB;
                break;
        }
        sumB += tmpSumB;
    });
    return sumB;
}
function getOpticalLiberationInLongitude(longitude, latitude, T) {
    var latRad = angleCalc_1.deg2rad(latitude);
    var i = 1.54242;
    var iRad = angleCalc_1.deg2rad(i);
    var phi = index_1.earthCalc.getNutationInLongitude(T);
    var F = getArgumentOfLatitude(T);
    var Omega = getMeanLongitudeOfAscendingNode(T);
    var W = angleCalc_1.normalizeAngle(longitude - phi - Omega);
    var WRad = angleCalc_1.deg2rad(W);
    var ARad = Math.atan2(Math.sin(WRad) * Math.cos(latRad) * Math.cos(iRad) - Math.sin(latRad) * Math.sin(iRad), Math.cos(WRad) * Math.cos(latRad));
    var A = angleCalc_1.normalizeAngle(angleCalc_1.rad2deg(ARad));
    return A - F;
}
exports.getOpticalLiberationInLongitude = getOpticalLiberationInLongitude;
function getOpticalLiberationInLatitude(longitude, latitude, T) {
    var latRad = angleCalc_1.deg2rad(latitude);
    var i = 1.54242;
    var iRad = angleCalc_1.deg2rad(i);
    var phi = index_1.earthCalc.getNutationInLongitude(T);
    var Omega = getMeanLongitudeOfAscendingNode(T);
    var W = angleCalc_1.normalizeAngle(longitude - phi - Omega);
    var WRad = angleCalc_1.deg2rad(W);
    var bRad = Math.asin(-1 * Math.sin(WRad) * Math.cos(latRad) * Math.sin(iRad) - Math.sin(latRad) * Math.cos(iRad));
    return angleCalc_1.rad2deg(bRad);
}
exports.getOpticalLiberationInLatitude = getOpticalLiberationInLatitude;
