// astronomy-bundle/planets/createPlanet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNeptune = exports.createUranus = exports.createSaturn = exports.createJupiter = exports.createMars = exports.createVenus = exports.createMercury = void 0;
var Mercury_1 = require("./Mercury");
var Venus_1 = require("./Venus");
var Mars_1 = require("./Mars");
var Jupiter_1 = require("./Jupiter");
var Saturn_1 = require("./Saturn");
var Uranus_1 = require("./Uranus");
var Neptune_1 = require("./Neptune");
function createMercury(toi) {
    return new Mercury_1.default(toi);
}
exports.createMercury = createMercury;
function createVenus(toi) {
    return new Venus_1.default(toi);
}
exports.createVenus = createVenus;
function createMars(toi) {
    return new Mars_1.default(toi);
}
exports.createMars = createMars;
function createJupiter(toi) {
    return new Jupiter_1.default(toi);
}
exports.createJupiter = createJupiter;
function createSaturn(toi) {
    return new Saturn_1.default(toi);
}
exports.createSaturn = createSaturn;
function createUranus(toi) {
    return new Uranus_1.default(toi);
}
exports.createUranus = createUranus;
function createNeptune(toi) {
    return new Neptune_1.default(toi);
}
exports.createNeptune = createNeptune;
